File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        
        if not subRoot or self.isSameTree(root, subRoot):
            return True

        left = self.isSubtree(root.left, subRoot)
        right = self.isSubtree(root.right, subRoot)

        return left or right
        

    def isSameTree(self, a: Optional[TreeNode], b: Optional[TreeNode]) -> bool:
        if not a and not b:
            return True

        if not a or not b or a.val != b.val:
            return False

        return self.isSameTree(a.left, b.left) and self.isSameTree(a.right, b.right)

File: Data Structures & Algorithms/same-binary-tree/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p, q) {
        const stack = [p, q]

        while (stack.length) {
            const qNode = stack.pop()
            const pNode = stack.pop()

            if (!pNode && !qNode) continue

            if ((!pNode && qNode) || (pNode && !qNode) || pNode.val != qNode.val) {
                return false
            }

            stack.push(pNode.right)
            stack.push(qNode.right)
            stack.push(pNode.left)
            stack.push(qNode.left)
        }
        return true
    }
}
